File: 0145-binary-tree-postorder-traversal/0145-binary-tree-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        if ( !root ){
              return { };
        }
         vector<int>Post; 
        /*(stack <TreeNode*> st;
        st.push(root);
         while ( !st.empty() ){
              TreeNode* node = st.top();
                  st.pop();
              if (node){
                  Post.push_back( node->val); 
              }
             if (node->left) {
                   st.push(node->left);
              } 
             if (node->right) {
                  st.push(node->right);
              }
         }
        reverse(Post.begin(),Post.end());
        */
        
         TreeNode * cur = root;
           while ( cur ) {
               
                if( cur->right ) {
                    
                     TreeNode * prece = cur->right;
                      
                     while ( prece->left && prece ->left != cur ){
                               prece = prece->left;
                     }
                      if ( prece->left) {
                          prece ->left = NULL ;
                           cur = cur->left;
                      }
                    else {
                        
                         Post.push_back(cur->val);
                          prece ->left = cur;
                          cur = cur->right;
                    }
                } else {
                        Post.push_back(cur->val);
                         cur = cur->left;
                }
           }
          reverse(Post.begin(), Post.end());
        return Post;
    }
};

File: 0145-binary-tree-postorder-traversal/0145-binary-tree-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
     Stack<TreeNode> st = new Stack<>();
     List<Integer>ans = new ArrayList<>();
      if(root == null)return ans;
      TreeNode cur = root;
      TreeNode lastVistied = null;
       while(cur != null || !st.isEmpty()){
               while(cur != null){
                    st.push(cur);
                    cur = cur.left;
               } 
               cur = st.peek();
               if(cur.right == null ||  lastVistied == cur.right){
                   ans.add(cur.val);
                   st.pop();
                lastVistied = cur;
                cur = null;
               } else{cur = cur.right;}  
       }
       return ans;
     /**
        root.left : 
        root.right :
         root.val :

       save root  :
          1
           2
              4


            4
            2 :  once popped measn check only rihg tside vlues:
                  
            1 


      */

    }
}